alf/utils/data_buffer.py

```diff
@@ -287,46 +296,62 @@ def dequeue(self, env_ids=None, blocking=False):
                 self._enqueued.wait(timeout=0.2)
             return None
         else:
-            return self._dequeue(env_ids)
+            return self._dequeue(env_ids=env_ids, n=n)
 
     @atomic
-    def _dequeue(self, env_ids=None):
-        """Return earliest batch and remove it from buffer.
+    def _dequeue(self, env_ids=None, n=1):
+        """Return earliest ``n`` steps and mark them removed in the buffer.
 
         Args:
             env_ids (Tensor): If None, ``batch_size`` must be num_environments.
                 If not None, dequeue from these environments. We assume there
                 is no duplicate ids in ``env_id``. ``result[i]`` will be from
                 environment env_ids[i].
+            n (int): Number of batches to dequeue.
         Returns:
-            nested Tensors of shapes ``[batch_size, ...]``.
+            nested Tensors of shapes ``[batch_size, n, ...]``.
         Raises:
             ``AssertionError`` when not enough data is present.
         """
         with alf.device(self._device):
             env_ids = self.check_convert_env_ids(env_ids)
             current_size = self._current_size[env_ids]
             min_size = current_size.min()
-            assert min_size >= 1, (
+            assert min_size >= n, (
                 "Not all environments have enough data. The smallest data "
                 "size is: %s Try storing more data before calling dequeue" %
                 min_size)
+            batch_size = env_ids.shape[0]
             pos = self._current_pos[env_ids] - current_size
             pos = pos % self._max_length
-            indices = env_ids * self._max_length + pos  # shape [B]
-
-            batch_size = env_ids.shape[0]
+            b_indices = env_ids.reshape(batch_size, 1).repeat(1, n)
+            t_range = torch.as_tensor([range(n)] * batch_size)
+            t_indices = (pos.reshape(batch_size, 1).repeat(1, n) +
+                         t_range) % self._max_length
             result = alf.nest.map_structure(
-                lambda buffer: buffer[indices].reshape(
-                    batch_size, 1, *buffer.shape[1:]), self._flattened_buffer)
-
-            self._current_size[env_ids] = current_size - 1
+                lambda buffer: buffer.__getitem__((b_indices, t_indices)).
+                reshape(batch_size, n, *buffer.shape[2:]),  # shape [B, n, ..]
+                self._buffer)
+            self._current_size[env_ids] = current_size - n
             # set flags if they exist to unblock potential consumers
             if self._dequeued:
                 self._dequeued.set()
                 self._enqueued.clear()
         return convert_device(result)
 
+    @atomic
+    def pop(self, n, env_ids=None):
```

Comment:
rename to `pop_up_to` or `dequeue_up_to`? otherwise it will be confusing

Reply:
good point.